src/hardware/mixer.cpp

```diff
@@ -368,6 +375,45 @@ void MixerChannel::AddSilence()
 	MIXER_UnlockAudioDevice();
 }
 
+void MixerChannel::EnableLowPassFilter(const bool enabled)
+{
+	filter.enabled = enabled;
+}
+
+void MixerChannel::ForceLowPassFilter(const bool force)
+{
+	filter.force = force;
+}
+
+void MixerChannel::ConfigureLowPassFilter(const uint8_t order,
+                                          const uint16_t cutoff_freq)
+{
+	assert(order > 0 && order <= max_filter_order);
+	const auto sample_rate = mixer.freq;
+	for (auto i = 0; i < 2; ++i)
+		filter.lpf[i].setup(order, sample_rate, cutoff_freq);
+}
+
+void MixerChannel::EnableZeroOrderHoldUpsampler(const bool enabled)
+{
+	zoh_upsampler.enabled = enabled;
+}
+
+void MixerChannel::UpdateZOHUpsamplerState()
+{
+	// we only allow upsampling
+	zoh_upsampler.step = std::min(static_cast<float>(sample_rate) /
+	                                      zoh_upsampler.target_freq,
+	                              1.0f);
+}
+
+void MixerChannel::ConfigureZeroOrderHoldUpsampler(const uint16_t target_freq)
```

Comment:
Is it possible for `target_freq` to exceed 65 Khz?  (my only concern is that some value larger than UINT16_MAX gets passed in and rolls over into the 16bit range.)

Reply:
Good point, the host rate might be 96k or 192k (not that it would make any sense to use those rates for playing DOS games, but well, technically very much a possibility).

Reply:
> It's only used for the SB which has the fixed DAC rate of 45454

Ahh perfect, in that case, let's leave it all as-is!